inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,172 @@
+                except_free_last_submissions.append(submission)
+            except:
+                pass
+
+        """Get the courses tasks, remove finished ones and courses that have no available unfinished tasks with upcoming deadline in range"""
+        tasks_data = {}
+        outdated_tasks=[]
+        for course in courses:
+            tasks = course.get_tasks()
+            for task in tasks:
+                the_task = course.get_task(task)
+                if (the_task.get_accessible_time().is_open()==False or ( (the_task.get_accessible_time().get_soft_end_date()) > (datetime.now()+timedelta(days=time_planner)) )): 
+                    outdated_tasks += [task]
+            new_user_task_list = course.get_task_dispenser().get_user_task_list([username])[username]
+            tasks_data.update({taskid: {"succeeded": False, "grade": 0.0} for taskid in new_user_task_list})
+
+            user_tasks = self.database.user_tasks.find({"username": username, "courseid": course.get_id(), "taskid": {"$in": new_user_task_list}})
+            for user_task in user_tasks:
+                tasks_data[user_task["taskid"]]["succeeded"] = user_task["succeeded"]
+                tasks_data[user_task["taskid"]]["grade"] = user_task["grade"]
+                if ( (tasks_data[user_task["taskid"]]["grade"] == 100.0)): 
+                    tasks_data.pop(user_task["taskid"])
+            "Remove outdated tasks and courses with no unfinished available tasks with deadline"""
+            for outdated_task in outdated_tasks:
+                if outdated_task in tasks_data:
+                    tasks_data.pop(outdated_task)
+            if (not any(task in tasks for task in tasks_data)):
+                open_courses.pop(course.get_id())
+
+        my_render=Render_Ordered(username)
+        time_planner = ["7", "14", "30", "unlimited"]
+        tasks_data_keys = list(tasks_data.keys())
+
+        """Sort the courses based on the most urgent task for each course"""
+        open_courses = OrderedDict( sorted(iter(open_courses.items()), key=lambda x: (sort_by_deadline(x[1], tasks_data_keys)[0]).get_accessible_time().get_soft_end_date() ))
+
+        return self.template_helper.render("coming_tasks.html",
+                                           template_folder=PATH_TO_PLUGIN + "/templates/",
+                                           open_courses=open_courses,
+                                           tasks_data=tasks_data,
+                                           tasks_list = list(tasks_data.keys()),
+                                           my_render=my_render,
+                                           time_planner=time_planner,
+                                           submissions=except_free_last_submissions)
+
+
+"""Class representing a render which is sent to the html file where it is recursively called with jinja"""
+class Render_Ordered:
```

Comment:
``RenderOrdered`` is more appropriate.

I do not understand well why you need to define this class as you only use it once, and the rendering method does not change.

You can simply use ``{% include %}`` jinja statements.

Reply:
Name changed.

Actually I probably created this class to counter my inexperience in Jinja...
I had difficulties to correctly setup the path for the html file (for recursion) within Jinja (PATH_TO_PLUGIN is defined in the python file and calling os.path.abspath(os.path.dirname(__file__)) inside Jinja is probably not a good practice either (plus it would require to import os). 
The second reason was also to easily call the specific order() method which is in the class. 
By creating a class, I was able to send an object containing the sorting method together with the plugin_path...
Probably not the bect practice but it was the easier way to implement this for me.

If you believe this is a real issue, let me know.

Reply:
Not an issue but I'm pretty sure this could be done in a easier way. It's ok as it is for me.

Reply:
Fine, I leave it like this even if it could probably be improved a bit.